### PR TITLE
Retry with no data

### PR DESCRIPTION
**Pull request by MatthewFlamm**
This ~partially~ mostly addresses #194.

This detects no data in observation and forecasts.  Also detects all stale forecasts if the right flag is passed.

### CODE REVIEW

**Comment by MatthewFlamm** [comment]
Need to add the keyword to all `update` methods.  It doesn't make a whole lot of sense to use this option for the alerts methods as often there will be no data, but it makes it easier to make a generic interface for all of them.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (MatthewFlamm) [inline .vscode/settings.json]: This needs to be moved I think

**Comment by MatthewFlamm** [comment]
This PR is ready for review.  I have implemented for the certainly most commonly used methods to retry if no data is returned.  The structure here is to require the `update_*` methods to accept a `raise_no_data` parameter that raises a custom `NwsNoDataError`.  This allows for an easy way for downstream applications to catch this without having to do the checking.  Then in `call_with_retry`, we use a `retry_no_data`, to automatically set this parameter to `True` and catch it for retries.  There is an edge case where both `retry_no_data` and `raise_no_data` are passed to `call_with_retry`.  This needs to be handled.

This is close enough for a meaningful review, so I will mark as ready for review.

Noticing codecov hasn't been commented lately.  Want to figure this out before merging this.

**Comment by lymanepp** [comment]
Until HA finally added humidity to weather forecasts, I was using pynws in a custom integration that I created. My integration had error handling and retry logic implemented as follows. It was an internal implementation in my pynws wrapper (which also supported other weather providers). It looked like this--I refactored this sample to make it part of pynws instead of a wrapper.

```
from functools import wraps
from typing import Any, Callable, Coroutine, ParamSpec, TypeVar


_P = ParamSpec("_P")
_T = TypeVar("_T")


class NwsError(Exception):
    """Error in Nws Class"""

    def __init__(self: NwsError, message: str, can_retry: bool = False):
        super().__init__(message)
        self.message = message
        self.can_retry = can_retry


class NwsNoDataError(NwsError):
    """No data was returned."""


def async_handle_exceptions(
    wrapped: Callable[_P, Coroutine[Any, Any, _T]]
) -> Callable[_P, Coroutine[Any, Any, _T]]:
    """`pynws` exception handler."""

    @wraps(wrapped)
    async def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> _T:
        try:
            return await wrapped(*args, **kwargs)
        except ClientResponseError as exc:
            raise NwsError("api_error", can_retry=exc.status >= 500) from exc
        except ClientError as exc:
            raise NwsError("cannot_connect", can_retry=True) from exc

    return wrapper


def async_retry(
    wrapped: Callable[_P, Coroutine[Any, Any, _T]]
) -> Callable[_P, Coroutine[Any, Any, _T]]:
    """`NwsError` retry handler."""

    @wraps(wrapped)
    async def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> _T:
        retries = 5
        while True:
            try:
                return await wrapped(*args, **kwargs)
            except NwsError as exc:
                if not exc.can_retry or retries == 0:
                    raise
                LOGGER.debug("%r: %d retries remaining", exc, retries)
                retries -= 1
            await asyncio.sleep(1)

    return wrapper



class SimpleNWS(Nws):
    ...

    @async_retry
    @async_handle_exceptions
    async def update_forecast(self: SimpleNWS) -> None:
        """Update forecast."""
        self._forecast = await self.get_gridpoints_forecast()
```

**Comment by MatthewFlamm** [comment]
Thanks for sharing!  This looks mostly similar to what has already been implemented, except we delegate to `tenacity` for the retry functionality.  Is there a piece you think should behave differently?

**Comment by MatthewFlamm** [comment]
Note that this PR adds in the ability for you to add in one more except block to your code, which I think shows that this PR adds in flexibility for your use case too.  But I want your feedback on that.  

PS I recommend using `tenacity` in your code if you want to simplify things.  I was recommended to add it here by a core dev to simplify the built-in NWS integration.  Even though I think it makes more sense to have this in HA directly, it does make some sense to have some sane default retry here as the NWS servers are so flaky.

**Comment by lymanepp** [comment]
> Is there a piece you think should behave differently?

You're right that the end result is pretty much the same. The main difference is becomes an implementation detail of `pwnws` instead of requiring that callers participate in the error handling and retries. It also hides the implementation details in decorators (AOP) instead of interspersing them into the "best-case" logic.

Edit: I'm a little OCD about "clean code". You probably already know that. 😉

**Comment by MatthewFlamm** [comment]
I'm not 100% sure I got your intent right.  You are proposing that the downstream apps should be the ones to handle the logic?  If I got this right, I would also prefer this, but HA core devs told me it would be required it to be here to get my proposed fix pushed :).  How I view this is `call_with_retry` provides some sane default behavior that is opt-in.  [Edit] But we should ensure that this doesn't muddle the non-retry built in behavior.  This was my purpose for making `tenacity` optional.

Again assuming I got your intent right, for downstream uses like you mentioned, the exposure of `raise_no_data` and `NwsNoDataError` in this PR should make it easier to catch this condition if you want.

[Second edit]  even thought I'm not a 100% fan of having the retry logic in `pynws`, it does make some small amount of sense since the NWS servers are so flaky, it will be a common use case.  If we can convince HA to adopt `tenacity` directly, we could consider providing a helper function here instead to aide in catching different error types.  Then we could deprecate + remove `call_with_retry`.

**Comment by lymanepp** [comment]
> You are proposing that the downstream apps should be the ones to handle the logic?

I'm proposing the opposite of that. The pynws library would handle all errors and retries internally (using the decorators). It could be enhanced to add `kwargs` to all public API methods so the caller could override or disable the default error handling and retries. But the default error handling and retries would be used by HA integration--so no need to override.

File: src/pynws/simple_nws.py [hunk 2]
```diff
@@ -52,23 +52,34 @@
 WIND: Final = {name: idx * 360 / 16 for idx, name in enumerate(WIND_DIRECTIONS)}
 
 
-def _is_500_error(error: BaseException) -> bool:
-    """Return True if error is ClientResponseError and has a 5xx status."""
-    return isinstance(error, ClientResponseError) and error.status >= 500
+def _nws_retry_func(retry_no_data):
+    def _retry(error: BaseException) -> bool:
+        """Return True if error is ClientResponseError and has a 5xx status."""
```
Comment (MatthewFlamm) [inline src/pynws/simple_nws.py]: Need better docstrings here to make logic clear.

File: tests/test_simple_nws.py [hunk 3]
```diff
@@ -275,6 +290,52 @@ async def test_nws_forecast_empty(aiohttp_client, mock_urls):
     assert forecast == []
 
 
+@freeze_time("2019-10-13T14:30:00-04:00")
+async def test_nws_forecast_empty_raise(aiohttp_client, mock_urls):
+    app = setup_app(gridpoints_forecast="gridpoints_forecast_empty.json")
+    client = await aiohttp_client(app)
+    nws = SimpleNWS(*LATLON, USERID, client)
+    with pytest.raises(NwsNoDataError):
```
Comment (MatthewFlamm) [inline tests/test_simple_nws.py]: Should add `match` condition to these

**Comment by lymanepp** [comment]
> . If we can convince HA to adopt `tenacity` directly, we could consider providing a helper function here instead to aide in catching different error types. Then we could deprecate + remove `call_with_retry`.

That would influence my thinking... As long as it was simple and unobtrusive. Sadly too much of HA is grotesquely complex and over-engineered (IMO). Edit: it's engineered for their full-time employees rather than the many contributors.

**Comment by MatthewFlamm** [comment]
> > You are proposing that the downstream apps should be the ones to handle the logic?
> 
> I'm proposing the opposite of that. The pynws library would handle all errors and retries internally (using the decorators). It could be enhanced to add `kwargs` to all public API methods so the caller could override or disable the default error handling and retries. But the default error handling and retries would be used by HA integration--so no need to override.

My rebuttal to this is that using `tenacity` directly is extremely flexible and I don't think we should replicate or try to wrap it's entire API inside pynws.  For example, what I'm suggesting is that if we make the `_nws_retry_func` (from this PR) public, then downstream apps could do something like this (it would need some naming thought before making public too).  This seems to be along the lines of what you are asking for, albeit not in pynws:


```
from pynws import nws_retry_func
from tenacity import retry, retry_if_exception, stop_after_delay, wait_fixed


class MyClass:

    @retry(
        reraise=True,
        wait=wait_fixed(10),
        stop=stop_after_delay(100),
        retry=retry_if_exception(nws_retry_func(retry_no_data=True)),
    )
    async def _update_observation(self):
        await self.nws.update_observation
```

**Comment by lymanepp** [comment]
> This seems to be along the lines of what you are asking for, albeit not in pynws:

I was just throwing it out there for consideration. I'll support whatever you choose to do!

**Comment by MatthewFlamm** [comment]
I want the other ideas as otherwise I could easily go off in a weird direction!  Thank you for suggesting them.  I also thought about having a one-line decorator that wrapped each update function like


```
class SimpleNWS:

    async def update_observation(...):

    @nws_retry
    async def update_observation_with_retry(...):
        await self.update_observation(...)
```

But then it becomes hard for end users to modify the retry behavior...

**Comment by lymanepp** [comment]
I was suggesting the following, to allow overriding the automatic retry behavior. The decorators have access to the `kwargs` and could define some standard keyword arguments to control their behavior.

```
class SimpleNWS:

    async def update_observation(...):

    @nws_retry
    async def update_observation_with_retry(..., **kwargs):
        await self.update_observation(...)
```

File: .vscode/settings.json (hunk 1 above)
Comment (MatthewFlamm) [reply .vscode/settings.json]: moved to .devcontainer

**Pull request merged**